### ISSUE TITLE
Add participants to public topics automatically when they engage (add idea, like, vote, etc)

### ISSUE THREAD

**Issue by ilmartyrk** [citizenos/citizenos-fe#307]
Currently any user can add arguments to public topic without being a member of that topic. In our redesign designs there is a text for users that are not members "Join this topic to engage in the discussion". It got me thinking that maybe we should add this extra step that users who want to join discussion should actually become members for the topic. 

Pros:
-  Users can get notifications about that topic and this topic will appear in their my topics list. Better overview for the owner of the topic about who are engaging their topic.
- Users can still always leave the topic if they wish
- Users will have to take a thinking moment before posting any arguments this can possibly avoid some spamming
Cons:
- Users will have to take an extra step to join
- For some topics user might be afraid to join to add their arguments

**Comment by anettlinno** [comment]
Triage 93. This needs more discussion. Sending it to In Prep.

**Comment by ilmartyrk** [comment]
@kevincrepin @BeccaMelhuish check if the issue is relevant or not.

**Comment by kevincrepin** [comment]
How I saw this was that there are two ways to "join" a topic:

1. You can join via the join button, which means this topic is now part of your feed and notifications but you didn't have to actively interact with it – you didn't have to join discussion or share ideas etc.
2. you can post an argument, leave a reply, share an idea or vote and this will automatically make you join the topic. 

So I didn't see this so that the user would not be able to interact with the topic without joining it. Interacting includes joining.
At least that's how I saw it :)

**Comment by BeccaMelhuish** [comment]
@kevincrepin totally agree - this makes a lot of sense to me. Anyone who has posted should be automatically turned into a participant, can't see any reason not to (or any reason to force them to actively click 'join' before they're allowed to post) :)

Will tag @ssin1901 though in case she sees something I'm missing :) 

Will also move this to soon as it does seem relevant to fix soon-ish

**Comment by BeccaMelhuish** [comment]
An example of this causing potential confusion/user issues in practice came to light with this one so I'll move it to 'Next?'. 

An umbrella organisation set their GA voting topic to public. As not everyone joined via a join link (presumably), they have more people having voted than they have registered participants: 

![image](https://github.com/user-attachments/assets/7fce1b3f-6835-406c-990b-78ce467768d9)

Anyone who votes should be automatically added as a participant, same as adding any other contribution (argument / idea / reply / reaction)

**Comment by ilmartyrk** [comment]
Okay, estimated dev time is 3h, will add the users as participants to the topic when posting idea, argument/reply or casting their vote.

My question is should we also do backward migration so that historic data would also change? This would take at least 8h to create a script that takes all the users that have done these actions in the past and are not members of the topics

**Comment by BeccaMelhuish** [comment]
@ilmartyrk Wonderful. Should we also add them if they react too? (I.e. like an idea or argument or reply). 

Re historic data, that's a good question. My general feeling is that we haven't had enough really active public topics in the past for it to be worthwhile, and also it's unlikely that old public topics are getting looked at much. 

But.. there are some currently or recently active public topics for which it would be great to have the correct participant numbers showing (IOF for example, plus some DD Foundation ones for which the issue might possibly be relevant). If we were to select a small number of recent/current public topics, would this make it any less time-consuming?

**Comment by ssin1901** [comment]
@BeccaMelhuish @ilmartyrk this fix hasn't been implemented in the test environment as the user added an idea but could not join the topic and could not take part in the voting either. Can I re-open?

![Screenshot 2024-11-25 at 00 04 59](https://github.com/user-attachments/assets/7bb6c2ec-ca1e-4b0f-b417-ca371965444e)

**Comment by BeccaMelhuish** [comment]
Yes please do just re-open and move back to 'in progress' if something isn't fixed :) 

(@ilmartyrk please contradict if not, but I think this is what we agreed)

**Comment by ilmartyrk** [comment]
@ssin1901 just to confirm did you test it in development.app.citizenos.com or in test.app.citizenos.com ? This fix is currently only in development.app.citizenos.com

**Comment by ssin1901** [comment]
@ilmartyrk ah, in test.app.citizenos.com. I'll test it in development.app.citizenos.com now

**Comment by ssin1901** [comment]
@ilmartyrk @BeccaMelhuish user added an idea but was not added to the topic (see screen recording)

https://github.com/user-attachments/assets/8ee7183e-6205-4837-af0e-d0006d874eee

**Comment by ssin1901** [comment]
Moving to in progress as fix is not implemented in developer environment yet

**Comment by BeccaMelhuish** [comment]
Also @ilmartyrk if it makes sense to fix both at once, then there is now this quite closely related issue too: https://github.com/citizenos/citizenos-fe/issues/1714

**Comment by ilmartyrk** [comment]
@BeccaMelhuish I would rather keep these issues apart

**Comment by ilmartyrk** [comment]
@ssin1901 @BeccaMelhuish the feature is actually working, but I guess not working was the part that the participants number was not updated, this will need some thinking how to properly fix

**Comment by BeccaMelhuish** [comment]
@ilmartyrk OK noted thank you, great that it's partially fixed already then! And as per the discussion in #1690, please also add people automatically to the topic when they turn on notifications :)

**Comment by BeccaMelhuish** [comment]
@ilmartyrk Tested this and it gave me a lot of joy to see it mostly working, but just a few tweaks still needed :) 

- [x] likes on ideas should make the user join
- [x] up or down reactions on arguments should make the user join
- [ ] 'join topic' should immediately switch to 'joined topic' when user turns on notifications (like it does for the others except idea adding as below)
- [ ] 'join topic' should immediately switch to 'joined topic' when user adds an idea 

Last two not super important if complex, but ideally this would happen.

**Comment by ilmartyrk** [comment]
Last two not simple to solve at the moment
- [x] likes on ideas should make the user join
- [x] up or down reactions on arguments should make the user join
- [ ] ~~'join topic' should immediately switch to 'joined topic' when user turns on notifications (like it does for the others except idea adding as below)~~
- [ ] ~~'join topic' should immediately switch to 'joined topic' when user adds an idea~~

**Comment by ssin1901** [comment]
Liking an idea (screen recording 1), adding an idea (screen recording 2) and upvoting an ideas (screen recording 3) automatically added the user to the topic, after a refresh

https://github.com/user-attachments/assets/725bbfa5-d0fa-4648-b8d9-1692e6113aeb
https://github.com/user-attachments/assets/8a457452-b07b-4461-b212-803dbfc7f432
https://github.com/user-attachments/assets/563aec3e-fcf9-4ee5-9552-e7cc5ca850a5

@BeccaMelhuish moving this to Pre-Live

**Comment by BeccaMelhuish** [comment]
@ilmartyrk @DiverVM @ssin1901 As per Slack message, important that this one doesn't get released until this one is fixed: 

https://github.com/citizenos/citizenos-fe/issues/1819
